dataland-community-manager/src/main/kotlin/org/dataland/datalandcommunitymanager/services/CompanyRolesManager.kt

```diff
@@ -0,0 +1,214 @@
+     */
+    @Transactional
+    fun assignCompanyRoleForCompanyToUser(
+        companyRole: CompanyRole,
+        companyId: String,
+        userId: String,
+    ): CompanyRoleAssignmentEntity {
+        val companyName = companyIdValidator.checkIfCompanyIdIsValidAndReturnName(companyId)
+        val correlationId = UUID.randomUUID().toString()
+
+        val companyRoleAssignmentEntityOptional = companyRoleAssignmentRepository.findById(
+            CompanyRoleAssignmentId(companyRole = companyRole, companyId = companyId, userId = userId),
+        )
+        return if (companyRoleAssignmentEntityOptional.isPresent) {
+            logCompanyRoleAlreadyAssigned(companyRole, companyId, userId)
+            companyRoleAssignmentEntityOptional.get()
+        } else {
+            val savedEntity = companyRoleAssignmentRepository.save(
+                CompanyRoleAssignmentEntity(companyRole = companyRole, companyId = companyId, userId = userId),
+            )
+            logCompanyRoleHasBeenAssigned(companyRole, companyId, userId)
+            if (companyRole == CompanyRole.CompanyOwner) {
+                companyOwnershipAcceptedEmailMessageSender.sendCompanyOwnershipAcceptanceExternalEmailMessage(
+                    userId, companyId, companyName, correlationId,
+                )
+            }
+            savedEntity
+        }
+    }
+
+    /**
+     * Logs that a user is already assigned to the specified company role for the company.
+     * @param companyRole that is already assigned
+     * @param companyId of the company for which the role is already assigned
+     * @param userId that is already assigned the company role to
+     */
+    private fun logCompanyRoleAlreadyAssigned(companyRole: CompanyRole, companyId: String, userId: String) {
+        logger.info(
+            "User with Id $userId already has the company role $companyRole for the company with " +
+                "the company Id $companyId",
+        )
+    }
+
+    /**
+     * Logs that a user has been assigned to the specified company role for the company.
+     * @param companyRole that has been assigned
+     * @param companyId of the company for which the role has been assigned
+     * @param userId that has been assigned the company role to
+     */
+    private fun logCompanyRoleHasBeenAssigned(companyRole: CompanyRole, companyId: String, userId: String) {
+        logger.info(
+            "User with Id $userId has received the company role $companyRole for the company with " +
+                "the company Id $companyId",
+        )
+    }
+
+    /**
+     * Returns company role assignments for the specified companyRole and company
+     * @param companyRole to filter for
+     * @param companyId for which the company role assignments are valid
+     * @returns the company role assignment entities
+     */
+    @Transactional
+    fun getCompanyRoleAssignmentsForRoleAndCompany(
```

Comment:
maybe we can add a function to get all role assignments for a company? Although this can be technically done by calling this function once for each role.

Also do we need a function to get all role assignments of a user in general? User X -> Role A in Company Y, Role B in company Z

Reply:
Yes you are right. The functions you are describing are pretty straight-forward to implement after my changes.
And I am also pretty sure, that we will need them soon.

However I have not implemented them yet, because they are not required in the current state of the code.
I did not want to do too much scope creep here.

We can add them in another PR, as soon as we need them.

Is that ok for you?

Reply:
absolutely fine